ntcore/util.py

```diff
@@ -0,0 +1,215 @@
+import threading
+import weakref
+
+from typing import Callable, Dict, Optional, Sequence
+
+from ._ntcore import NetworkTableInstance, NetworkTableEntry, Value
+
+__all__ = ["ntproperty", "ChooserControl"]
+
+NetworkTables = NetworkTableInstance.getDefault()
+
+
+class _NtProperty:
+    """
+    Don't use this directly, use @ntproperty instead
+    """
+
+    ntvalue: NetworkTableEntry
+
+    _instlock = threading.Lock()
+    _instances: Dict[int, weakref.WeakSet] = {}
+
+    @classmethod
+    def attach(cls, self: "_NtProperty", inst: NetworkTableInstance):
+        with cls._instlock:
+            handle = inst._getHandle()
+            props = cls._instances.get(handle)
+            if not props:
+                props = weakref.WeakSet()
+                cls._instances[handle] = props
+            props.add(self)
+
+    @classmethod
+    def onInstancePostReset(cls, inst: NetworkTableInstance):
+        with cls._instlock:
+            props = cls._instances.get(inst._getHandle(), [])
+            for prop in props:
+                prop.reset()
+
+    @classmethod
+    def onInstanceDestroy(cls, inst: NetworkTableInstance):
+        with cls._instlock:
+            cls._instances.pop(inst._getHandle(), None)
+
+    def __init__(
+        self,
+        key: str,
+        defaultValue,
+        writeDefault: bool,
+        persistent: bool,
+        inst: NetworkTableInstance,
+    ) -> None:
+        self.key = key
+        self.defaultValue = defaultValue
+        self.writeDefault = writeDefault
+        self.persistent = persistent
+        # never overwrite persistent values with defaults
+        if persistent:
+            self.writeDefault = False
+        self.inst = inst
+        _NtProperty.attach(self, inst)
+        self.reset()
+
+    def reset(self):
+        self.ntvalue = self.inst.getEntry(self.key)
+        if self.writeDefault:
+            self.ntvalue.setValue(self.defaultValue)
+        else:
+            self.ntvalue.setDefaultValue(self.defaultValue)
+
+        if self.persistent:
+            self.ntvalue.setPersistent()
+
+        # this is an optimization, but presumes the value type never changes
+        self.mkv = Value.getFactoryByType(self.ntvalue.getType())
+
+    def get(self, _):
+        return self.ntvalue.value
+
+    def set(self, _, value):
+        self.ntvalue.setValue(self.mkv(value))
+
+
+def ntproperty(
+    key: str,
+    defaultValue,
+    writeDefault: bool = True,
+    doc: str = None,
+    persistent: bool = False,
+    *,
```

Comment:
Since this is technically a new API, should we make these kw-only?

Reply:
Good point. Going to leave key/defaultValue as positional though.